### PR TITLE
[Native] Fix bf16 squared distance for near-duplicate vectors

### PR DESCRIPTION
**Pull request by flobernd**
## Summary

The AVX-512 BF16 and SVE bf16 x bf16 squared-distance kernels compute `|a-b|^2` as `a.a + b.b - 2a.b` using bf16
dot products. Every term is rounded to f32 at the magnitude of `a.a + b.b`, so for near-duplicate vectors the
difference is dominated by rounding noise and can turn negative. Lucene's distance-to-score mapping requires a
non-negative distance, so a negative result yields a wrong or infinite score.

The kernels keep the dot-product form. When a result falls below `(a.a + b.b) / 4096`, the kernel recomputes that
vector with the exact base kernel, which computes `(a-b)^2` directly in f32. The base kernel signature is declared
in a shared header so the higher-capability kernels can call it. The fallback is entirely in native, so callers
receive an exact, non-negative squared distance with no change on their side.

## Changes

- Recompute near-duplicate bf16 x bf16 squared distances in the AVX-512 and SVE kernels by calling the exact base
  kernel, declared in a shared header.
- Add a test covering exact and near-duplicate vectors at unnormalised scale for the single, bulk, offsets and
  sparse entry points.
- Add a near-duplicate fraction parameter to the bf16 bulk-operation benchmark so the fallback cost can be
  measured as the clash rate rises.
- Bump the vec native library to 1.0.146.

## Related

Supersedes #158817.

## Notes

The recompute runs only for the near-duplicate vectors a kernel flags; every other input takes the vectorized
fast path unchanged. Measured on Sapphire Rapids, Zen 5 and Graviton 4, the in-kernel recompute is within noise of
the fast path at realistic clash rates and the zero-duplicate path is unchanged.

### CODE REVIEW

**Comment by thecoop** [comment]
I'm currently running a test to determine how often clashes occur in our datasets.

Initial thoughts - I think simdvec is the wrong abstraction that this should live. I would prefer a layer above that, maybe at the scorer/vectorizationprovider layer, so that it could gracefully degrade from native -> panama -> scalar. That way we could recalculate using the panama code we've already got. Single-score would return -1 indicating recalculate, bulk would return a value indicating if any scores need to be recalculated (maybe a bitset on a `long` based on a max bulk size of 64? Or just a non-zero value, and then the scores array is scanned for -1).

However, the architecture needed for that may be prohibitive, or not make sense based on what we've currently got.

We also need to check what performance impact this has at the index level for a variety of our datasets

**Comment by ldematte** [comment]
> Initial thoughts - I think simdvec is the wrong abstraction that this should live. I would prefer a layer above that, maybe at the scorer/vectorizationprovider layer, so that it could gracefully degrade from native -> panama -> scalar.

++
And we already have a similar pattern, using `Float.NEGATIVE_INFINITY` to indicate "this scorer did not run/did not give you a valid score" (e.g. `MemorySegmentScorer#scoreBulk`). Or a boolean in case we do not need to return a value. We could/should have the same here I think.

**Comment by ldematte** [comment]
E.g. looking at org/elasticsearch/simdvec/internal/BFloat16VectorScorer.java:99, we could: 
1. change `dotProductDBF16QF32BulkSparse` to return a boolean, not void, record it, and use alongside "resolved" to fall back to another bulk impl.
2. keep `dotProductDBF16QF32BulkSparse` signature the same, read the flags in the loop, recompute what needed (even just the "wrong" scores, using a fallback single-score impl).
3. do _also_ the check Java side: after all, we already loop over results there.

My preference goes to 2 (and possibly -inf as a flag for consistency with the rest of the codebase); I think the problem detection belongs to the native code, because it depends on that impl, but the fallback belongs to the scorer (BFloat16VectorScorer in this case).

**Comment by thecoop** [comment]
I've checked glove, wiki-cohere, ann-gist, hotpotqa for duplicates and clashes when rounded to bf16.

Glove contains no clashes.
wiki has 99 identical pairs and 35 very similar pairs, involving 223 out of 934,000 vectors.
ann-gist has 32,368 identical pairs and 730 very similar pairs involving 33,491 vectors out of 1,000,000 vectors.
hotpotqa-E5-small has 4,746,545 identical pairs involving 15,118 out of 7,000,000 vectors - there's lots of repeats in there.

So it's definitely a minority, ~~1% of all vectors have clashes of some kind across a variety of datasets.

One option is to check for identical vectors in the native code, and return the correct result in that case (which would catch most of them), and then have the fallback case for very close clashes. That check might be expensive for the common case though.

As for the fallback, if we're falling back to panama, that would need to live at the scorer level, as that's the level that knows about different native/panama implementations. But if we're falling back to the caps_1 native implementation, that _is_ something that could live at the simdvec library level, as it's just involving native code implementations. That might be cleaner (and faster to run).

Unless there's a way to call the caps_1 code directly from the caps_2 code, maybe using headers for the references? Then it can be entirely in native...

**Comment by flobernd** [comment]
Thanks for the data backed analysis @thecoop !

Let me quickly do some performance benchmarks with a panama fallback implementation on scorer level to decide whether this might be sufficient or whether we rather want something faster, like the caps_1 fallback.

**Comment by thecoop** [comment]
OK, hows this for a plan:

- Declare the caps_1 bf16-bf16 method signatures in a header so they can be referenced from the caps_2 code
- Check the clash condition at the end of the caps_2 bf16-bf16 methods
- For all vectors that clash, recalculate them using the caps_1 code (which uses f32 diff - diff rather than dot-product). For the bulk methods, we can use stack-allocated arrays for offsets/addresses when calling the caps_1 method, so we don't need to malloc anything. We'll only be recalculating the clashing vectors in a batch.
- Profit!

This means we're not having to maintain two different caps_2 implementations whilst keeping performance, and this is all entirely native changes, Java is completely unaware of this.

We definitely need to do this for SVE; whether we change the AVX code to be better on intel but worse for zen or also use fallbacks depends on how expensive the extra check-and-fallback turns out to be.

I don't think an inline identical-vector check is feasible, we would need to maintain more register data to keep an XOR + OR acc of each vector bits with the query bits to check for identicalness, that'll cause register spillage and significantly more instructions for the common case.

**Comment by flobernd** [comment]
@thecoop Sounds good. I'll play around with this idea and also benchmark it on all platforms.

File: libs/simdvec/native/src/vec/c/aarch64/vec_bf16_2.cpp [hunk 1]
```diff
@@ -395,21 +398,32 @@ static inline void sqrDbf16Qbf16_bulk_sve(
...
     }
 
     // vectors tail
     for (; c < count; c++) {
         const bf16_t* a0 = mapper(a, c, offsets, pitch);
         results[c] = vec_sqrDbf16Qbf16_2(a0, b, dims);
     }
+
+    // Recompute the near-duplicate vectors marked above (results[i] < 0) with batched base-kernel calls.
+    sqr_bf16_recompute_clashes<TData, mapper>(a, b, dims, pitch, offsets, count, results);
```
Comment (thecoop) [inline libs/simdvec/native/src/vec/c/aarch64/vec_bf16_2.cpp]: Can we condition this on whether any of the `flag_clash` methods return -1? As we get back each modified value, we just need to check if it's -1, then set a bool we need to recompute something

File: libs/simdvec/native/src/vec/headers/vec_bf16_1.h [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    const f32_t result,
+    const bf16_t* a,
+    const bf16_t* b,
+    const int32_t elementCount
+) {
+    return result < self * SQR_BF16_CANCELLATION_GUARD ? vec_sqrDbf16Qbf16(a, b, elementCount) : result;
+}
+
+// Batch size for recomputing bulk clashes through stack-allocated arrays, so nothing is malloc'd.
+static constexpr int32_t SQR_BF16_CLASH_BATCH = 64;
```
Comment (thecoop) [inline libs/simdvec/native/src/vec/headers/vec_bf16_1.h]: 32 may be a more sensible value, that's the default batch size for a lot of the bulk calls IIRC

File: libs/simdvec/native/src/vec/headers/vec_common.h [hunk 3]
```diff
@@ -55,6 +55,23 @@ inline auto sqr_scalar<uint16_t, f32_t>(const uint16_t a, const f32_t b) {
...
+ * below this fraction of a.a + b.b would have more than roughly 2^-12 relative error and can even
+ * be negative, so the higher-capability kernels recompute it with the exact base kernel (see
+ * vec_bf16_1.h).
+ */
+static constexpr f32_t SQR_BF16_CANCELLATION_GUARD = 1.0f / 4096.0f;
+
+// Marks `result` as a clash (returns -1) when it fell below the guard relative to `self` (a.a + b.b), so a
+// later pass can recompute it exactly with the base kernel; otherwise returns `result` unchanged. Used by the
+// bulk kernels, where the clashing vectors are recomputed together in one batched call (see vec_bf16_1.h).
+static inline f32_t sqr_bf16_flag_clash(const f32_t self, const f32_t result) {
```
Comment (thecoop) [inline libs/simdvec/native/src/vec/headers/vec_common.h]: These should be in the bf16 header

**Comment by ldematte** [comment]
>Declare the caps_1 bf16-bf16 method signatures in a header so they can be referenced from the caps_2 code
Check the clash condition at the end of the caps_2 bf16-bf16 methods
For all vectors that clash, recalculate them using the caps_1 code (which uses f32 diff - diff rather than dot-product). For the bulk methods, we can use stack-allocated arrays for offsets/addresses when calling the caps_1 method, so we don't need to malloc anything. We'll only be recalculating the clashing vectors in a batch.
Profit!

I like it; we can definitely reference caps_1 methods from caps_2, just a header or even just a declaration stuck on top of the caps_2 file will do. The header is cleaner IMO.

> we can use stack-allocated arrays for offsets/addresses when calling the caps_1 method

That would be nice; but for a fallback like this we could even just call the caps_1 single scorer I think. I would avoid malloc if we can.

**Comment by thecoop** [comment]
> That would be nice; but for a fallback like this we could even just call the caps_1 single scorer I think. I would avoid malloc if we can.

True, maybe the single-scorer is enough for the fixup

**Comment by flobernd** [comment]
I used the single scorer initially, but pivoted to bulk now as suggested earlier. I'm currently running benchmarks. Let's see which version wins 🙂

**Comment by flobernd** [comment]
Single form is performance equivalent to batch + bool-gate. I'll go with the single scorer since it introduces less complexity.

File: libs/simdvec/src/test/java/org/elasticsearch/simdvec/SimdVecLibraryBFloat16Tests.java [hunk 4]
```diff
@@ -312,6 +313,110 @@ public void testBFloat16BulkSparseScattered() {
...
+        // single
+        for (int v = 0; v < numVecs; v++) {
+            float expected = ScalarOperations.similarity(function, base, docs[v]);
+            var docSeg = bf16Segment.asSlice((long) v * dims * BFloat16.BYTES, (long) dims * BFloat16.BYTES);
+            assertDistance("single vector " + v, expected, similarity(docSeg, querySeg, dims));
+        }
+
+        // bulk (sequential)
+        int[] identity = new int[numVecs];
+        for (int v = 0; v < numVecs; v++) {
```
Comment (thecoop) [inline libs/simdvec/src/test/java/org/elasticsearch/simdvec/SimdVecLibraryBFloat16Tests.java]: `Arrays.setAll(IntUnaryOperator.identity())`

File: libs/simdvec/src/test/java/org/elasticsearch/simdvec/SimdVecLibraryBFloat16Tests.java [hunk 5]
```diff
@@ -312,6 +313,110 @@ public void testBFloat16BulkSparseScattered() {
...
+        float[] expected = new float[numVecs];
+        ScalarOperations.bulkWithOffsets(function, base, docs, identity, expected);
+        var scores = arena.allocate((long) numVecs * Float.BYTES);
+        similarityBulk(bf16Segment, querySeg, dims, numVecs, scores);
+        assertDistances("bulk", expected, scores);
+
+        // bulk with offsets (random ordinals)
+        int[] ordinals = new int[numVecs];
+        for (int v = 0; v < numVecs; v++) {
+            ordinals[v] = randomInt(numVecs - 1);
```
Comment (thecoop) [inline libs/simdvec/src/test/java/org/elasticsearch/simdvec/SimdVecLibraryBFloat16Tests.java:382]: Does it matter we could get duplicates here? Could have an identity array and then shuffle it

Comment (flobernd) [reply libs/simdvec/src/test/java/org/elasticsearch/simdvec/SimdVecLibraryBFloat16Tests.java:382]: Duplicates don't affect correctness here since the expected scores are built from the same ordinals, so each position is checked against the same document; they just let a vector be scored more than once in one bulk call, which is realistic for HNSW neighbor lists, whereas a shuffle would guarantee distinct offsets but drop that coverage, so I'd keep it as-is.

**Comment by thecoop** [comment]
I'm really happy with where this has ended up.

I think the only thing missing is a benchmark - could we add a benchmark parameter to the BF16 suites that injects a variable number of near-duplicates into the benchmark data? Maybe ranging between 0 and 10%?

**Comment by ldematte** [comment]
> I'm really happy with where this has ended up.

++, it's nice and self contained, correct but still able to extract the best performance.
Besides the benchmark @thecoop mentioned, please remember to update the PR description.

**Comment by flobernd** [comment]
Great, thanks to both of you for refining this with me. Also very happy with the end result 🎉

I'll add the benchmark parameter and update the PR description again.

**Review by thecoop — APPROVED** [review]
I think this is great. Shipit!
